clang/lib/AST/PreorderAST.cpp

```diff
@@ -377,6 +377,127 @@ void PreorderAST::NormalizeExprsWithoutConst(Node *N) {
   AddNode(L, B);
 }
 
+bool PreorderAST::GetDerefOffset(Node *UpperNode, Node *DerefNode,
```

Comment:
It seems to me that this function could be simplified.  You could have one function that checks whether two leaf nodes are integer constant expressions and computes their difference if any.   This function could just walk children looking for two integer constant expressions.

I thought one of the points of normalization is to use associativity and commutivity to group nodes together, i.e. `a + (b + c)` and `(a + b) + c` are put into normal forms that match.   Another point is that it groups constant expressions together so that they can be constant-folded into a single expression.  This should then simplify this code.

Reply:
Yes, this is true if we only had the addition operator. Then the tree would be flattened with a single BinaryNode (+) and all children sorted and the integers constant-folded. Then we could simply walk the tree matching expressions and compute the difference only for the constant expressions.

The reason I had to recursively match was to support multiplication. Consider two expressions like `a + (b * 2) + 1` and `a + (b * 2) + 2`. In this case we cannot coalesce the two BinaryNodes and the two trees are:
```
+                              +
 \                              \
  a, 1, *                        a, 2, *
          \                              \
           b, 2                           b, 2
```

With that said, I think I need to slightly change my logic to compute the difference only for integer constant expressions which are children of the addition operator and ensure all children of a multiplication operator are equal.